Add cube map texture support.

rongguodong opened the pull request:
cubemap_shader uses the cube map texture in "sky box" mode.
cubemap_reflection_shader uses the cube map texture in reflection mode.

The new sample "gvr-cubemap" demonstrates both modes.

GearVRf-DCO-1.0-Signed-off-by: Guodong Rong
rongguodong@hotmail.com

rongguodong commented:
Since loading a cube map texture asynchronously is not trivial, the new sample uses synchronous version. It is just to show how to use a cube map texture in both sky box mode and reflection mode. We can create asynchronous loading for cube map texture later, and modify the sample accordingly.

chenchao1407 commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -489,6 +493,22 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
+        }
+
+        return resourceList.size() != 6 ? null : new GVRCubemapTexture(this,
+                bitmapList);
```
I think the size check should be at the beginning other wise it may be a performance waste when the texture is not valid.

chenchao1407 commented on GVRf/Framework/src/org/gearvrf/GVRMaterial.java:
```diff
@@ -97,7 +97,17 @@
+            public static final GVRMaterialShaderId ID = new GVRStockMaterialShaderId(
+                    7);
+        }
+};
```
These two classes should be inside GVRShaderType.

chenchao1407 commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -341,6 +341,10 @@ public GVRMesh createQuad(float width, float height) {
+        mesh.setNormals(normals);
+
         float[] texCoords = { 0.0f, 0.0f, 0.0f, 1.0f, 1.0f, 0.0f, 1.0f, 1.0f };
         mesh.setTexCoords(texCoords);
```
Just my personal opinion, since these normals and texCoords are constant values. I think we'd better to make them as final static member of this class.

rongguodong commented:
I have modified code according to Chao and Jon's comments. Particularly, I added asynchronous loading support for cube map texture and use asynchronous loading in the new sample now.

I have also fixed bugs in old asynchronous loading code, and rebased with Tom's change.

Please review and test the code again. Thanks!

rongguodong commented:
Note on the sample code:
There are two sets of bitmaps (beach.zip and screenshot3d.zip). The current code uses beach.zip for cube map texture since it is more beautiful. The commented out code uses screenshot3d.zip and uses the setFaceNames method.

chenchao1407 commented on GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java:
```diff
@@ -0,0 +1,135 @@
+                ZipEntry ze = null;
+                while ((ze = zipInputStream.getNextEntry()) != null) {
+                    String iamgeName = ze.getName();
+                    Integer imageIndex = GVRCubemapTexture.faceIndexMap
```
I think it should be imageName.

rongguodong replied:
Fixed.

JonShemitz commented on GVRf/Framework/jni/objects/textures/cubemap_texture.h:
```diff
@@ -0,0 +1,93 @@
+namespace gvr {
+class CubemapTexture: public Texture {
+public:
+	explicit CubemapTexture(JNIEnv* env, jobjectArray bitmapArray) :
```
Everywhere else, the `_jni.cpp` files contain all the JNI code, and make calls into the 'straight C++ code'. I think it would be best to follow that convention, here: do all the `env->` in `cubemap_texture_jni.cpp` and just pass in an array of six structs with the `AndroidBitmapInfo info` and the `void* pixels`.

rongguodong replied:
I moved env->GetArrayLength(bitmapArray) to cubemap_texture_jni.h.

For env->GetObjectArrayElement and AndroidBitmap_* functions, I think it is more efficient to be in cubemap_texture.h. If we move all of them to cubemap_texture_jni.h, we need to create six objects there and pass them to cubemap_texture.h, and delete the six objects then. This will introduce necessary memory allocation and delocation. I think it is the reason why base_texture.h has all AndroidBitmap_* functions in it, not in base_texture_jni.h.

JonShemitz replied:
> If we move all of them to cubemap_texture_jni.h, we need to create six objects there and pass them to cubemap_texture.h, and delete the six objects then. This will introduce necessary memory allocation and delocation. 

Allocating and deallocating a local array of `struct`s is very cheap ,,, but the existence of this grody code in `base_texture.cpp` sort of undercuts my Follow The Format argument, here.

JonShemitz commented on GVRf/Framework/jni/shaders/material/cubemap_reflection_shader.cpp:
```diff
@@ -0,0 +1,182 @@
+// is at bottom left, when we use Android bitmap to create OpenGL texture, it is already
+// up-side down. So we do not need to flip them again.
+//
+// We do need to flip the z-coordinate to be consistent with the left-handed system.
```
Nice!

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
+     * 
+     * <p>
+     * Note that this method may take hundreds of milliseconds to return: unless
+     * the bitmap is quite tiny, you probably don't want to call this directly
```
`bitmap` s/b `cube map`

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
+        assertGLThread();
+
+        if (resourceArray.length != 6) {
+            return null;
```
Should throw an `IllegalArgumentException`

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
+
+        Bitmap[] bitmapArray = new Bitmap[6];
+        for (int i = 0; i < 6; i++) {
+            bitmapArray[i] = GVRAsynchronousResourceLoader.decodeStream(
```
I probably ought to move `decodeStream()` to the utility package.

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRContext.java:
```diff
@@ -1093,6 +1136,29 @@ public void loadTexture(TextureCallback callback,
+     *            six images are "posx.png", "negx.png", "posy.png", "negx.png",
+     *            "posz.png", and "negz.png", which can be changed by calling
+     *            {@link GVRCubemapTexture#setFaceNames(String[])}.
+     * @returnA {@link Future} that you can pass to methods like
```
`@returnA` s/b `@return A`

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java:
```diff
@@ -0,0 +1,83 @@
+/** Cube map texture. */
+public class GVRCubemapTexture extends GVRTexture {
+    /**
+     * Constructs a cube map texture using a pre-existing {@link Bitmap}s.
```
`a` s/b `6`

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java:
```diff
@@ -0,0 +1,83 @@
+     *            map texture respectively.
+     */
+    public static void setFaceNames(String[] nameArray) {
+        if (nameArray.length != 6) {
```
Should be using Chao's `Preconditions` (aka `Assert`) class.

JonShemitz commented on GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java:
```diff
@@ -0,0 +1,83 @@
+     * Internal map data structure. Public only for visible to
+     * {@code AsyncCubemapTexture}.
+     */
+    public static Map<String, Integer> faceIndexMap = new HashMap<String, Integer>(
```
* Should probably be `final` 
* Is there a way to make it `private` (or package-visible) and explicitly pass it to `AsyncCubemapTexture`?

JonShemitz commented on GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java:
```diff
@@ -0,0 +1,135 @@
+
+    private static final String TAG = Log.tag(AsyncCubemapTexture.class);
+
+    private static final Class<? extends GVRHybridObject> TEXTURE_CLASS = GVRTexture.class;
```
I'm pretty sure this breaks asynch textures - you need to register `AsyncCubemapTexture` here.

JonShemitz commented on GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java:
```diff
@@ -0,0 +1,135 @@
+                    Integer imageIndex = GVRCubemapTexture.faceIndexMap
+                            .get(imageName);
+                    if (imageIndex == null) {
+                        throw new IllegalArgumentException("Name of image ("
```
The `Exceptions` class would let you use `String.format()` syntax.

JonShemitz commented on GVRf/Framework/src/org/gearvrf/asynchronous/GVRAsynchronousResourceLoader.java:
```diff
@@ -299,6 +301,44 @@ public void run() {
+     * @param priority
+     *            This request's priority. Please see the notes on asynchronous
+     *            priorities in the <a href="package-summary.html#async">package
+     *            description</a>. Also, please note priorities only apply to
```
Zap `Also, please note ... request scheduler.`

JonShemitz commented:
Looks really good. Aside from
1. The JNI code outside of a `_jnp.cpp` class and
2. The probable mis-registration in `AsyncCubemapTexture.java`
all my comments are really pretty minor.

rongguodong commented:
I pushed a new commit addressing Jon's comments. For condition checking parts, since we are modifying the Assert class, I can modify them after #44 are merged.

chenchao1407 commented on GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java:
```diff
@@ -0,0 +1,137 @@
+            Bitmap[] bitmapArray = new Bitmap[6];
+            ZipInputStream zipInputStream = new ZipInputStream(
+                    resource.getStream());
+
```
Just a minor concern here.

Forcing developers to use zip to construct a cubetexture may lose some flexibility for some cases e.x. : if one want to use several combinations of same bunch of textures.

JonShemitz replied:
> several combinations of same bunch of textures.

Perhaps you and Guodong discussed this in-person? From what I know of cube maps, this doesn't seem like a very likely scenario: while I'm sure you could come up with examples (manufactured objects with pre-fab, snap-in panels?) I gather that most cube maps are renders (or captures) of complex objects; most of the time, each face will be a unique texture. I may (of course!) be wrong.

chenchao1407 commented:
Just reviewed the codes. Looks fine to me aside from some minor concerns.

Tested. Runs well. Good to go merged if there is no any other objection.

Do you have any other comment? @JonShemitz
